# Targeted-PYR1-Library-Design-App

Targeted-PYR1-Library-Design-App is an App that allows for the generation of oligos to make PYR1 protein libraries using general screen datasets to inform those libraries. 

## Using the app online

If you want to use the app online simply go to [https://weslgeorge.shinyapps.io/project-oysters/](https://weslgeorge.shinyapps.io/project-oysters/) shinyapps.io link and get started! 


## Getting setup locally:

If you want to use the app locally on your computer simply clone the repository into your local files. Make sure to have R and Rstudio installed on your computer and open the app.R file using Rstudio.

Then install the required packages with the following code in the console.

```{install packages}
install.packages("shiny", "shinyjs", "doParallel", "spatstat.utils", "bioseq", "DT","tidyverse","openxlsx")
```

Then click on Run in the top right in Rstudio or by using the following code in your R console while in the cloned directory.

```{Run app}
shiny::runApp()
```

The app should open on the Readme tab and you can get started on generating PYR1 libraries!

## Using the app

The general workflow for designing to generate PYR1 mutant libraries is as follows. A sequence profile can first be generated in the app or manually uploaded to the app. The sequence profile is just a set of substitution mutations allowed for oligo design, and wild type amino acids will always be included regardless of if they are specified in the profile. Following the upload or creation of the sequence profile, the app will break the PYR1 coding sequence into three distinct oligo blocks; block 1 being from nucleic acid position 153 to 306; block 2 being from nucleic acid position 303 to 449; and block 3 being from nucleic acid position 446 to 572. Due to this the sequence profile will only affect amino acids from amino acid position K59 to Q189 which is sufficient to mutate the PYR1 binding pocket. These oligo blocks are then appended with flanking barcoding sequences to allow for amplification, and internal BsaI restriction enzyme cut sites that allow cloning with the pBD-PYR1-156bp-F1 plasmid backbone to be used in yeast-two-hybrid chemical screens in combination with MAV99 harboring both pACT-HAB1 and the plasmid library constructed. 

### Sequence profile generation

  First target small molecule interactors for your future library are identified, and similarly structured compounds are found in the general screen data. This can be done with external chemical clustering methods, or by using programs like [Chemminetools](https://chemminetools.ucr.edu/) to cluster a target compound with the general screen data SMILES. Identified similar small molecules used to inform the library will be selected in the `Sequence profile generation` tab to generate a sequence profile that informs the app what mutations at what positions in PYR1 to allow in the library. Selecting desired small molecules will filter for biosensor mutations that were identified in biosensors for the selected small molecules that then populate the sequence profile. Sequence profiles can be made manually and uploaded in the sidebar using the `Sequence profile CSV` upload button, as well as made in the `Sequence profile generation` tab. See the `Sequence profile` sheet of the `Oligo summary data` download button for a template of the file, new rows can be added as long as the proper WT amino acid and position are added. 

### Library parameter selection

  Next use the sidebar to inform the parameters for the library. Input a library name to help organize oligos in a larger Twist order, then select what PYR1 scaffold you wish to use and the number of mutations per block you wish to allow for. (note that increasing the number of mutations per block may make the app slow for a time while it generates oligos). The constitutive filter will filter out single and double mutation site blocks with known constitutive activity including two key residues that are commonly constitutive, and can help reduce library size. Navigate to the `Oligo summary information` tab to see details on library size, and expected number of PYR1s generated by the library, and oligos needed for its construction. As a reference we will normally aim to make libraries between 10,000 and 500,000 unique PYR1s to have a realistically screenable library size. 

### Barcoding oligo blocks

  The third step is to use the `Oligo barcoding` tab to select the barcodes and cut sites you want to use to flank each oligo block. This allows you to make multiple libraries and order them with a single oligo pool order from Twist biosciences, cutting down on library costs. Typically each block in a library gets a unique oligo barcode which we have defaulted to “set####” primers which we use in our own lab. These primer sequences will be in the downloadable report, and additionally can be browsed in the `pop_experiment.xlsx - 10K_primers.csv` file (Not all of these primers have been tested). Barcodes can also be entered manually if your lab has existing primers commonly used. Our library approach clones into the pBD-PYR1-153bp-F1 backbone using golden gate assembly; other cutsites can be selected but only BsaI cut sites will work with this backbone.    

### Downloading files 

  Finally make sure to download both the `Oligos.fasta` file and `Oligo summary data` file in the from the sidebar once all tabs have been opened and the library is to your liking. The oligos fasta file will generate a list of oligos that can be submitted to Twist bioscience orders and have the naming convention of >LIBRARY\_`input library name`\_PYR1\_`PYR1 type`\_FRAG\_`Block number`\_mutation1:mutation2:mutation3:mutation4_PRI\_`barcode primers`. The `Oligo summary data` will download an excel file that contains sheets including the sequence profile used to generate the library, oligo summary information, and a table of the oligos and the barcodes used. 

## License & DOI

This project is licensed under the GNU General Public License, version 3 (GPLv3) - see the LICENSE file for details, and the DOI for the app is [![DOI](https://zenodo.org/badge/969746437.svg)](https://doi.org/10.5281/zenodo.15319591).  

## App Citations 

We used R version 4.4.3 (R Core Team 2025) and the following R packages: bioseq v. 0.1.4 (Keck 2020), doParallel v. 1.0.17 (Corporation and Weston 2022), DT v. 0.33 (Xie, Cheng, and Tan 2024), openxlsx v. 4.2.8 (Schauberger and Walker 2025), shiny v. 1.10.0 (Chang et al. 2024), shinyjs v. 2.1.0 (Attali 2021), shinythemes v. 1.2.0 (Chang 2021), spatstat.utils v. 3.1.3 (Baddeley, Turner, and Rubak 2025), tidyverse v. 2.0.0 (Wickham et al. 2019).

Attali, Dean. 2021. shinyjs: Easily Improve the User Experience of Your Shiny Apps in Seconds. https://CRAN.R-project.org/package=shinyjs.

Baddeley, Adrian, Rolf Turner, and Ege Rubak. 2025. spatstat.utils: Utility Functions for “spatstat”. https://CRAN.R-project.org/package=spatstat.utils.

Chang, Winston. 2021. shinythemes: Themes for Shiny. https://CRAN.R-project.org/package=shinythemes.

Chang, Winston, Joe Cheng, JJ Allaire, Carson Sievert, Barret Schloerke, Yihui Xie, Jeff Allen, Jonathan McPherson, Alan Dipert, and Barbara Borges. 2024. shiny: Web Application Framework for r. https://CRAN.R-project.org/package=shiny.

Corporation, Microsoft, and Steve Weston. 2022. doParallel: Foreach Parallel Adaptor for the “parallel” Package. https://CRAN.R-project.org/package=doParallel.

Keck, Francois. 2020. “Handling Biological Sequences in r with the Bioseq Package.” Methods in Ecology and Evolution. https://doi.org/10.1111/2041-210X.13490.

R Core Team. 2025. R: A Language and Environment for Statistical Computing. Vienna, Austria: R Foundation for Statistical Computing. https://www.R-project.org/.

Schauberger, Philipp, and Alexander Walker. 2025. openxlsx: Read, Write and Edit Xlsx Files. https://CRAN.R-project.org/package=openxlsx.

Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy D’Agostino McGowan, Romain François, Garrett Grolemund, et al. 2019. “Welcome to the tidyverse.” Journal of Open Source Software 4 (43): 1686. https://doi.org/10.21105/joss.01686.

Xie, Yihui, Joe Cheng, and Xianying Tan. 2024. DT: A Wrapper of the JavaScript Library “DataTables”. https://CRAN.R-project.org/package=DT.


Grateful was used to generate citations for this app

Rodriguez-Sanchez F, Jackson C (2024). _grateful: Facilitate citation of R packages_.
  <https://pakillo.github.io/grateful/>.

  





